fat_fs: Add support for the "win" variable alignment in FATFS struct

ngboonkhai opened the pull request:
This is to support Sypnosys Designware SDMMC controller read and write
operation where the buffer address is needed to be 16bytes and 512bytes
aligned. Adding macro FS_FATFS_WINDOW_ALIGNMENT to align the "win" variable
address in FATFS struct.

Signed-off-by: Boon Khai Ng <boon.khai.ng@intel.com>

ngboonkhai committed:
fat_fs: Add support for the "win" variable alignment in FATFS struct

This is to support Sypnosys Designware SDMMC controller read and write
operation where the buffer address is needed to be 16bytes and 512bytes
aligned. Adding macro FS_FATFS_WINDOW_ALIGNMENT to align the "win"
variable address in FATFS struct.

Signed-off-by: Boon Khai Ng <boon.khai.ng@intel.com>

de-nordic reviewed (changes requested):
The change requires west manifest update that should be part of the PR.

ngboonkhai commented:
> The change requires west manifest update that should be part of the PR.

Hi Ermel,

Thanks for reviewing my PR, should I go and update the west manifest? If yes, can point me the documentation on what should I update on the west manifest. Thanks.

Regards,
Boon Khai

de-nordic commented:
> Thanks for reviewing my PR, should I go and update the west manifest? If yes, can point me the documentation on what should I update on the west manifest. Thanks.

This change applies only when certain changes to FAT FS are applied (https://github.com/zephyrproject-rtos/fatfs/pull/9), so what you need is to have commit that would propose the change to the FAT FS for by providing modified west.yml that would refer to the PR you want to introduce and then the commit that adds Zephyr changes on top.

This is my PR https://github.com/zephyrproject-rtos/zephyr/pull/36302, as an example, that has been introducing new version of FAT FS into Zephyr; there are two commits: the first one introduces new version of FAT FS by changing its manifest sha and the other commit adds some supporting code. This is very similar to your case here.
The difference is that you need to start with west.yml that, instead of containing sha, will refer to your pull request to fatfs PR as: `pull/9/head`

ngboonkhai commented:
> > Thanks for reviewing my PR, should I go and update the west manifest? If yes, can point me the documentation on what should I update on the west manifest. Thanks.
> 
> This change applies only when certain changes to FAT FS are applied ([zephyrproject-rtos/fatfs#9](https://github.com/zephyrproject-rtos/fatfs/pull/9)), so what you need is to have commit that would propose the change to the FAT FS for by providing modified west.yml that would refer to the PR you want to introduce and then the commit that adds Zephyr changes on top.
> 
> This is my PR #36302, as an example, that has been introducing new version of FAT FS into Zephyr; there are two commits: the first one introduces new version of FAT FS by changing its manifest sha and the other commit adds some supporting code. This is very similar to your case here. The difference is that you need to start with west.yml that, instead of containing sha, will refer to your pull request to fatfs PR as: `pull/9/head`

Hi Ermel,

Ah, okay got it.

So for the  PR in ([zephyrproject-rtos/fatfs#9](https://github.com/zephyrproject-rtos/fatfs/pull/9)) I still need to retain right?
then for this PR, I will need to resubmit with a force push with a updated west manifest?

Regards,
Boon Khai

de-nordic commented:
> So for the PR in ([zephyrproject-rtos/fatfs#9](https://github.com/zephyrproject-rtos/fatfs/pull/9)) I still need to retain right? then for this PR, I will need to resubmit with a force push with a updated west manifest?
> 

Yes, you need both PRs because this PR, the one we are talking under, is actually attempting to introduce, into Zephyr, changes that you have done to the other module via the https://github.com/zephyrproject-rtos/fatfs/pull/9.
This allows anyone to use west command with your PR to download and test your solution.
When you will have the https://github.com/zephyrproject-rtos/fatfs/pull/9 accepted, then once again you will have to update this PR with the sha that your merged changes will get on the FAT FS module tree.

ngboonkhai commented:
> > So for the PR in ([zephyrproject-rtos/fatfs#9](https://github.com/zephyrproject-rtos/fatfs/pull/9)) I still need to retain right? then for this PR, I will need to resubmit with a force push with a updated west manifest?
> 
> Yes, you need both PRs because this PR, the one we are talking under, is actually attempting to introduce, into Zephyr, changes that you have done to the other module via the [zephyrproject-rtos/fatfs#9](https://github.com/zephyrproject-rtos/fatfs/pull/9). This allows anyone to use west command with your PR to download and test your solution. When you will have the [zephyrproject-rtos/fatfs#9](https://github.com/zephyrproject-rtos/fatfs/pull/9) accepted, then once again you will have to update this PR with the sha that your merged changes will get on the FAT FS module tree.

Hi Ermel,

I found that there is no reviewer in the pull request [zephyrproject-rtos/fatfs#9](https://github.com/zephyrproject-rtos/fatfs/pull/9) . I cant find any button to add the reviewer, can you help to review my code at the PR as well?

Regards,
Boon Khai

nvlsianpu reviewed (commented):
@ngboonkhai You should add reference to the fatfs module PR in west.yaml patch.
Something like in https://github.com/zephyrproject-rtos/zephyr/pull/41262/files

ngboonkhai commented:
> @ngboonkhai You should add reference to the fatfs module PR in west.yaml patch. Something like in https://github.com/zephyrproject-rtos/zephyr/pull/41262/files

Hi nvlsianpu,

Thanks for remind, Ermel has told me that I need to update the west.yaml as well, was waiting for the PR at [zephyrproject-rtos/fatfs#9.](https://github.com/zephyrproject-rtos/fatfs/pull/9)

After that I will force push the PR here with the hash with the updated west.yaml.

Wonder if you can help me the PR at https://github.com/zephyrproject-rtos/fatfs/pull/9 ?

Regards,
Boon Khai

nvlsianpu commented:
@ngboonkhai contribution process expects that you should first have PR like this one with `pull/9/head` reference for fatfs module - this allows us to see CI results. Then once ( with approvals on this PR and the fatfs module PR) fatfs PR will be merged that reference will needed to be changed to the SHA.

de-nordic commented on subsys/fs/Kconfig.fatfs:
```diff
@@ -129,6 +129,13 @@ config FS_FATFS_MAX_SS
 
+config FS_FATFS_WIN_ALIGNED
+	int "Memory alignment for the member \"win\" in FATFS"
+	default 0
```
I am not sure whether 0 is correct alignment, shouldn't be 1 as default? There is no point to align anything to 0.

ngboonkhai replied:
Hi Ermel,

Yup, it shouldn't align with 0 although I use the 0 as a switch to detect for :no alignment", which is not a good implementation. Will default the value to 1 and resubmit the PR.

de-nordic commented on subsys/fs/Kconfig.fatfs:
```diff
@@ -129,6 +129,13 @@ config FS_FATFS_MAX_SS
+	default 0
+	help
+	  Specify the alignment in byte eg: 16, 512. This is a low-level
+	  requirement by some of the mmc controller.
```
The alignment happens inside FATFS structure which  means that the size of structure may significantly get bigger, when for example 512 alignment is selected. I think this should be noted here.

ngboonkhai replied:
Hi Ermel,

Okay will add the note to indicate that. Thanks.

de-nordic commented on subsys/fs/Kconfig.fatfs:
```diff
@@ -129,6 +129,13 @@ config FS_FATFS_MAX_SS
 	range 512 4096
 	default 512
 
+config FS_FATFS_WIN_ALIGNED
```
Rename the same way it is now called in https://github.com/zephyrproject-rtos/fatfs/blob/32f6befc3d4dda481f9649663e5122196d2b7c06/include/ffconf.h#L257

ngboonkhai replied:
Hi @de-nordic 

Done force pushed.

Regards
Boon Khai.

de-nordic commented:
> > @ngboonkhai You should add reference to the fatfs module PR in west.yaml patch. Something like in https://github.com/zephyrproject-rtos/zephyr/pull/41262/files
> 
> Hi nvlsianpu,
> 
> Thanks for remind, Ermel has told me that I need to update the west.yaml as well, was waiting for the PR at [zephyrproject-rtos/fatfs#9.](https://github.com/zephyrproject-rtos/fatfs/pull/9)
> 
> After that I will force push the PR here with the hash with the updated west.yaml.
> 
> Wonder if you can help me the PR at [zephyrproject-rtos/fatfs#9](https://github.com/zephyrproject-rtos/fatfs/pull/9) ?
> 
> Regards, Boon Khai

Can  you modify the west.yml to pull in the FAT FS PR? Otherwise the sanity here is run with FAT FS that does not support the option you are adding.
Your PR here should consist of two commits:
 1) the one that modifies the west.yml to include changes in the FAT FS pr;
 2) the additions that make the changes from 1) available to FAT FS driver within Zephyr.

The FAT FS PR will not be merged till this PR is able to run with it included through west. 
When sanity completes with the PR reference, we can merge the FAT FS PR and then you will have to, again, modify the west.yml to, this time, include sha of the FAT FS commit that would be including your change to FAT FS repo.
Then, if everything goes OK, we can merge this PR.

ngboonkhai commented:
> > > @ngboonkhai You should add reference to the fatfs module PR in west.yaml patch. Something like in https://github.com/zephyrproject-rtos/zephyr/pull/41262/files
> > 
> > 
> > Hi nvlsianpu,
> > Thanks for remind, Ermel has told me that I need to update the west.yaml as well, was waiting for the PR at [zephyrproject-rtos/fatfs#9.](https://github.com/zephyrproject-rtos/fatfs/pull/9)
> > After that I will force push the PR here with the hash with the updated west.yaml.
> > Wonder if you can help me the PR at [zephyrproject-rtos/fatfs#9](https://github.com/zephyrproject-rtos/fatfs/pull/9) ?
> > Regards, Boon Khai
> 
> Can you modify the west.yml to pull in the FAT FS PR? Otherwise the sanity here is run with FAT FS that does not support the option you are adding. Your PR here should consist of two commits:
> 
> 1. the one that modifies the west.yml to include changes in the FAT FS pr;
> 2. the additions that make the changes from 1) available to FAT FS driver within Zephyr.
> 
> The FAT FS PR will not be merged till this PR is able to run with it included through west. When sanity completes with the PR reference, we can merge the FAT FS PR and then you will have to, again, modify the west.yml to, this time, include sha of the FAT FS commit that would be including your change to FAT FS repo. Then, if everything goes OK, we can merge this PR.
> 
> Btw. I did test builds of this change, after checking out all needed repos by hand (because west is not done...) and here is example of memory impact, for your information. with window set to 1 (default):
> 
> ```
> west build -d ttt -b nrf52840_blip zephyr/samples/subsys/fs/fat_fs/ -- -DCONFIG_FS_FATFS_WINDOW_ALIGNMENT=1
> ...
> Memory region         Used Size  Region Size  %age Used
>            FLASH:       47320 B         1 MB      4.51%
>             SRAM:        8952 B       256 KB      3.41%
>         IDT_LIST:          0 GB         2 KB      0.00%
> ```
> 
> and with window set to 512:
> 
> ```
> west build -d ttt -b nrf52840_blip zephyr/samples/subsys/fs/fat_fs/ -- -DCONFIG_FS_FATFS_WINDOW_ALIGNMENT=512
> ...
> Memory region         Used Size  Region Size  %age Used
>            FLASH:       47336 B         1 MB      4.51%
>             SRAM:        9824 B       256 KB      3.75%
>         IDT_LIST:          0 GB         2 KB      0.00%
> ```

Hi @de-nordic , 

Thanks for helping to run the compilation manually.

I don't quite understand on this part, if the PR in the fatfs not merge in yet, how do I get the revision id?
The only id can get is this 32f6befc3d4dda481f9649663e5122196d2b7c06. However when it try to clone the fatfs manually, i cant checkout to the commit id.

Regards,
Boon Khai

de-nordic commented:
> 
> I don't quite understand on this part, if the PR in the fatfs not merge in yet, how do I get the revision id? The only id can get is this 32f6befc3d4dda481f9649663e5122196d2b7c06. However when I try to clone the fatfs manually, i cant checkout to the commit id.
> 

You use the /pull/9/head as the sha:
```
...
    - name: fatfs
      revision: /pull/9/head
      path: modules/fs/fatfs
      groups:
...
```

This allows sanity, and interested reviewers, to get your complete solution and test it.

ngboonkhai committed:
west.yml: update fatfs for new config option

Update west.yaml to the commit consist of the new config option
CONFIG_FATFS_WINDOW_ALIGNMENT to align the win variable with
specific alignment

Signed-off-by: Boon Khai Ng <boon.khai.ng@intel.com>

ngboonkhai commented:
> > I don't quite understand on this part, if the PR in the fatfs not merge in yet, how do I get the revision id? The only id can get is this 32f6befc3d4dda481f9649663e5122196d2b7c06. However when I try to clone the fatfs manually, i cant checkout to the commit id.
> 
> You use the [/pull/9](https://github.com/zephyrproject-rtos/zephyr/pull/9)/head as the sha:
> 
> ```
> ...
>     - name: fatfs
>       revision: /pull/9/head
>       path: modules/fs/fatfs
>       groups:
> ...
> ```
> 
> This allows sanity, and interested reviewers, to get your complete solution and test it.



> > I don't quite understand on this part, if the PR in the fatfs not merge in yet, how do I get the revision id? The only id can get is this 32f6befc3d4dda481f9649663e5122196d2b7c06. However when I try to clone the fatfs manually, i cant checkout to the commit id.
> 
> You use the [/pull/9](https://github.com/zephyrproject-rtos/zephyr/pull/9)/head as the sha:
> 
> ```
> ...
>     - name: fatfs
>       revision: /pull/9/head
>       path: modules/fs/fatfs
>       groups:
> ...
> ```
> 
> This allows sanity, and interested reviewers, to get your complete solution and test it.

Hi @de-nordic ,

Okay, done added. 
Thanks.

But after the review process, I need to force push again with the revision id that merged in into the fatfs?

Regards,
Boon Khai

ngboonkhai commented:
> > I don't quite understand on this part, if the PR in the fatfs not merge in yet, how do I get the revision id? The only id can get is this 32f6befc3d4dda481f9649663e5122196d2b7c06. However when I try to clone the fatfs manually, i cant checkout to the commit id.
> 
> You use the [/pull/9](https://github.com/zephyrproject-rtos/zephyr/pull/9)/head as the sha:
> 
> ```
> ...
>     - name: fatfs
>       revision: /pull/9/head
>       path: modules/fs/fatfs
>       groups:
> ...
> ```
> 
> This allows sanity, and interested reviewers, to get your complete solution and test it.

Hi @de-nordic 

The /pull/9/head doesn't seems to work.

Regards,
Boon Khai.

de-nordic commented on west.yml:
```diff
@@ -42,7 +42,7 @@ manifest:
         - tools
     - name: fatfs
-      revision: 94fcd6bfb3801ac0a5e12ea2f52187e0a688b90e
+      revision: /pull/9/head
```
Sorry, my fault: should be `pull/9/head`, without leading `/`.

ngboonkhai replied:
Hi @de-nordic 

Ah okay, no issue :)  force pushed with the changes without leading / .

Regards, 
Boon Khai

de-nordic commented:
> 
> But after the review process, I need to force push again with the revision id that merged in into the fatfs?
> 

Yest, that will ave to happen.

de-nordic commented:
The process is described here: https://docs.zephyrproject.org/latest/guides/modules.html#changes-to-existing-module

ngboonkhai commented:
> > But after the review process, I need to force push again with the revision id that merged in into the fatfs?
> 
> Yest, that will ave to happen.

Ah okay

ngboonkhai commented:
> The process is described here: https://docs.zephyrproject.org/latest/guides/modules.html#changes-to-existing-module

Okay, noted.

de-nordic reviewed (changes requested):
Change order of commits: west change goes in first.

In west commit remove `pull/9/head` from commit title and body, because the same data will be within source anyway, instead  mention what the update actually brings in, as a rationale for the update; look at commit 284fbf0bf38452a6fb2a6474795c757c8c15bc1a, as an example, where you will see that the commit message describes what the west update brings in with module.

ngboonkhai commented on subsys/fs/Kconfig.fatfs:
```diff
@@ -129,6 +129,16 @@ config FS_FATFS_MAX_SS
+	default 1
+	help
+	  Specify the alignment in byte eg: 16, 512. This is a low-level
+	  requirement by some of the mmc controller. The alignement is
```
Updated with the correct spelling

ngboonkhai commented:
> Change order of commits: west change goes in first.
> 
> In west commit remove `pull/9/head` from commit title and body, because the same data will be within source anyway, instead mention what the update actually brings in, as a rationale for the update; look at commit [284fbf0](https://github.com/zephyrproject-rtos/zephyr/commit/284fbf0bf38452a6fb2a6474795c757c8c15bc1a), as an example, where you will see that the commit message describes what the west update brings in with module.

Hi Ermel,

Alright, done force push with the the above request 

Regards,
Boon Khai

nvlsianpu commented:
@ngboonkhai you can left behind west update now - it was bumped up by #41885

ngboonkhai commented:
> @ngboonkhai you can left behind west update now - it was bumped up by #41885

Hi @nvlsianpu,

Meaning i need to rebase my repo to the latest and force push?

Regards,
Boon Khai

nvlsianpu commented:
yes

ngboonkhai commented:
> yes

@nvlsianpu how do I rerun the twister without force push? the twister failed, due to I update the zephyr mainline first before update the fatfs repo, and now I need to rerun the twister.

de-nordic commented:
@ngboonkhai I have merged the FAT FS module, please update revision in west.yaml

ngboonkhai commented on subsys/fs/Kconfig.fatfs:
```diff
@@ -129,6 +129,16 @@ config FS_FATFS_MAX_SS
+	int "Memory alignment for the member \"win\" in FATFS"
+	default 1
+	help
+	  Specify the alignment in byte eg: 16, 512. This is a low-level
```
Updated the patch with suggested help information

de-nordic reviewed (approved):
Looks OK. Thanks!

ngboonkhai commented:
Hi @de-nordic  @nvlsianpu  

Please help me to merge the changes. 

Thanks.

de-nordic commented:
> Please help me to merge the changes.
> 

Unfortunately you have to wait till code-freeze ends for the 3.0 release (https://github.com/zephyrproject-rtos/zephyr/wiki/Program-Management#actual-and-planned-milestone-dates), then we will be poking at people to merge it.
